### ISSUE TITLE
Eloquent WhereIn working incorrect when models collection is bypassed.

### ISSUE THREAD

**Issue by a-komarev** [laravel/framework#16808]
- Laravel Version: 5.3.26
- PHP Version: 7.0.7
- Database Driver & Version: MySQL Ver 14.14 Distrib 5.6.28

### Description:

When trying to bypass `get()` to `whereIn` method instead of `pluck('id')` - query seems to be broken and no errors are thrown.

### Steps To Reproduce:

```php
$statuses = Status::whereIn('name', [
    'draft',
    'pending_approval',
    'awaiting_payment',
    'paid',
    'paid_part',
    'another_status',
])->get();

$payments = Payment::whereIn('status_id', $statuses)->get();
```

This code generates query:

```sql
select *
  from `payments`
 where `status_id` in ('3', 'awaiting_payment', '1', 'draft', '5', 'another_status');
```

Expected output:

```sql
select *
  from `payments`
 where `status_id` in ('3', '2', '1', '4', '5', '6');
```

*If `pluck('id')` will be bypassed - everything work as expected.*

**Comment by lagbox** [comment]
"When trying to bypass not pluck('id') of collection, but get() to whereIn method "  can you rephrase that? And what does 'trying to bypass not ...' mean?

**Comment by lasselehtinen** [comment]
I am not sure why you would want to bypass it anyway. Now you are just generating unnecessary work for the database to try to find values that are not there.

But even more importantly, imagine if you had relationships in the original statuses table with a different index or a normal int column . Then those values would be also searched as a possible value for the status_id, which you do not want. So in general you should be explicit when defining the values in whereIn. Adding the pluck is a minor deal IMHO.

**Comment by a-komarev** [comment]
@lagbox rephrased.

@lasselehtinen I've never tried to drop full collection of items to `whereIn` method before. Usually I'm using `pluck()`, transforming\mapping it to array of ids, using joins or `whereHas` on relation:

```php
$payments = Payment::whereHas('status', function (Builder $query) {
    $query->whereIn('name', [
        'draft',
        'pending_approval',
        'awaiting_payment',
        'paid',
        'paid_part',
        'another_status',
    ]);
})->get();
```

I've dropped full collection while wrote a quick draft of the future code and saw that not all the expected results was displayed. After debugging output SQL I found that `whereIn` used each key of the collection as value to find because under the hood it's just converting collection to the array and adds each key to the where statement. Now I see that there is no way to *magically* resolve what keys should be used for query and there is no way to make `get()` work exactly the same way as `pluck()` in `whereIn` in query builder.
